### PR TITLE
Fixed #23408 -- Added migrations questioner prompt for adding unique fields with a callable default.

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
ticket-23408

### CODE REVIEW

File: django/db/migrations/state.py [hunk 1]
```diff
@@ -670,6 +671,14 @@ def __init__(self, app_label, name, fields, options=None, bases=None, managers=N
                     'ModelState.fields cannot refer to a model class - "%s.through" does. '
                     'Use a string reference instead.' % name
                 )
+            # Warn if a callable default is provided for fields having unique=True.
+            if field.unique and field.default is not NOT_PROVIDED and callable(field.default):
+                warnings.warn(
+                    'Callable defaults will not generate unique values upon migrating. '
+                    f"Found callable default {field.default} on unique field '{self.name}.{name}'. "
+                    'See https://docs.djangoproject.com/en/dev/howto/writing-migrations/'
```
Comment (charettes) [inline django/db/migrations/state.py:679]: I supposed we'd want to link to either `stable` or the current `django.VERSION` instead of `dev`.

**Review by charettes — COMMENTED** [review]
It feels like this logic should be added to the auto-detector/questioner level instead

https://github.com/django/django/blob/d400b08a8baa697905daadd47a6ba12336e93336/django/db/migrations/autodetector.py#L869-L882

https://github.com/django/django/blob/d400b08a8baa697905daadd47a6ba12336e93336/django/db/migrations/questioner.py#L108-L164

e.g. by having the auto-detector ask to continue or abort if `field.unique and field.default is not NOT_PROVIDED and callable(field.default)` instead of emitting a warning

File: tests/migrations/test_state.py [hunk 2]
```diff
@@ -1016,6 +1018,23 @@ class Meta:
...
+        class Author(models.Model):
+            random_seed = models.FloatField(unique=True, default=random.Random)
+
+            class Meta:
+                app_label = 'migrations'
+                apps = new_apps
+
+        msg = 'Callable defaults will not generate unique values upon migrating. '
+        msg += "Found callable default <class 'random.Random'> on unique field 'Author.random_seed'. "
+        msg += 'See https://docs.djangoproject.com/en/dev/howto/writing-migrations/#migrations-that-add-unique-fields'
```
Comment (smithdc1) [inline tests/migrations/test_state.py:1033]: Something like this should work? 

```suggestion
        msg = (
            'Callable defaults will not generate unique values upon migrating. '
            "Found callable default <class 'random.Random'> on unique field 'Author.random_seed'. "
            'See https://docs.djangoproject.com/en/dev/howto/writing-migrations/#migrations-that-add-unique-fields'
        )
```

File: django/db/migrations/state.py (hunk 1 above)
Comment (jacobtylerwalls) [reply django/db/migrations/state.py:679]: Ah! We have `get_docs_version()`.

**Comment by jacobtylerwalls** [comment]
> It feels like this logic should be added to the auto-detector/questioner level instead

Great, I wondered that myself but wanted to get input before moving the ticket away from warning to interactive questioner. Happy to retool it.

File: tests/migrations/test_commands.py [hunk 3]
```diff
@@ -1870,6 +1872,64 @@ class Meta:
...
+        class Book(models.Model):
+            created = models.DateTimeField(unique=True, default=timezone.now)
+
+            class Meta:
+                app_label = 'migrations'
+
+        version = get_docs_version()
+        input_msg = (
+            'Callable defaults will not generate unique values upon migrating. '
+            'Found callable default on unique field book.created.\n'
```
Comment (kezabelle) [inline tests/migrations/test_commands.py:1890]: `book` should be `Book` ideally, I think?

I wonder also if the `Found ...` sentence should first. Telling the user about callable defaults without having told them what was found seems stiff. Though if it's following the cadence of other prompts (I can't recall), then disregard; better to be consistent.

Comment (jacobtylerwalls) [reply tests/migrations/test_commands.py:1890]: Good ear! I sort of adapted the existing cadence and made it worse by making it two complete sentences. I'll spruce up into one.

Comment (jacobtylerwalls) [reply tests/migrations/test_commands.py]: About caps, yah spotted that also, but I noticed status quo, see:

https://github.com/django/django/blob/604df4e0adc71da264f61fe85020a170c98e6f09/tests/migrations/test_commands.py#L1375-L1378

Do you think we should open an easy pickings ticket to capitalize the model in each affected prompt?

**Review by charettes — APPROVED** [review]
Thanks for the changes Jacob,

Changes and tests look great to me!

File: tests/migrations/test_commands.py (hunk 3 above)
Comment (felixxm) [reply tests/migrations/test_commands.py]: I don't think it's worth changing :shrug:

**Comment by felixxm** [comment]
@jacobtylerwalls Thanks :+1: 

I pushed minor edits.

**Pull request merged**